webpack/webpack-embedded-dependencies-plugin/src/index.ts

```diff
@@ -0,0 +1,178 @@
+      normalModuleFactory.hooks.module.tap(
+        PLUGIN_NAME,
+        (module, moduleCreateData: IWebpackModuleCreateData, resolveData) => {
+          const { resourceResolveData } = moduleCreateData;
+          const pkg: IPackageData | undefined = resourceResolveData?.descriptionFileData;
+          const filePath: string | undefined = resourceResolveData?.descriptionFileRoot;
+
+          if (pkg && filePath && filePath.includes('node_modules')) {
+            const key: PackageMapKey = makePackageMapKeyForPackage(pkg);
+            thirdPartyPackages.set(key, { dir: filePath, data: pkg });
+          }
+
+          return module;
+        }
+      );
+    });
+
+    // Tap into compilation so we can tap into compilation.hooks.processAssets
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tapAsync(
+        { name: PLUGIN_NAME, stage: Compilation.PROCESS_ASSETS_STAGE_REPORT },
+        async (assets, callback) => {
+          const rawPackages: FlattenedPackageEntry[] = Array.from(thirdPartyPackages).sort((first, next) => {
+            return first[0].localeCompare(next[0]);
+          });
+
+          const packages: IPackageData[] = [];
+
+          for (const [, { dir, data }] of rawPackages) {
+            const { name, version } = data;
+            const license: string | undefined = parseLicense(data);
+            const copyright: string | undefined = (await parseCopyright(dir)) || parsePackageAuthor(data);
+            packages.push({
+              name,
+              version,
+              license,
+              copyright
+            });
+          }
+
+          const dataToStringify: IEmbeddedDependenciesFile = {
+            embeddedDependencies: packages
+          };
+
+          console.log(dataToStringify);
+
+          compilation.emitAsset(this.outputFileName, new sources.RawSource(JSON.stringify(dataToStringify)));
+
+          callback();
+        }
+      );
+    });
+  }
+}
+
+function makePackageMapKeyForPackage(pkg: IPackageData): PackageMapKey {
+  return `${pkg.name}@${pkg.version}`;
+}
+
+/**
+ * Returns the license type
+ */
+function parseLicense(packageData: IPackageData): string | undefined {
+  if (packageData.license) {
+    return packageData.license;
+  } else if (typeof packageData.licenses === 'string') {
+    return packageData.licenses;
+  } else if (packageData.licenses && packageData.licenses.length) {
+    return packageData.licenses.length === 1
+      ? packageData.licenses[0].type
+      : `(${packageData.licenses
+          .map((license: { type: string; url: string }) => license.type)
+          .join(' OR ')})`;
+  }
+
+  return undefined;
+}
+
+/**
+ * Given a module path, try to parse the module's copyright attribution.
+ */
+async function parseCopyright(modulePath: string): Promise<string | undefined> {
+  // Add copyright info
+  const licenseFile: string[] = await globAsync(path.join(modulePath, LICENSE_FILES));
```

Comment:
Pass `{ fs: compiler.inputFileSystem }` in the options structure to `globAsync`. Either `fs` or `memfs` implement all the necessary methods.

Reply:
Also, using `path.join` on a glob specifier is not semantically valid, because `\` is a reserved character in a glob. You need to use `cwd: modulePath` in the options instead.

Reply:
> Pass `{ fs: compiler.inputFileSystem }` in the options structure to `globAsync`. Either `fs` or `memfs` implement all the necessary methods.

`glob` doesn't seem to have a `fs` property for its options in its types however I do see the API listed in the documentation. I'll give this a try and see how far I get.

Reply:
> Another note, since you're only looking top-level in the folder, you might be better served throwing out `glob` altogether and just issuing a single call to `compiler.inputFileSystem.readdir` and sifting through the results via a Set of allowed file names.

I chose this path but will definitely want a re-review on the implementation.